### PR TITLE
fix(config/html): handle encoding issues and improve error handling in config and HTML file loading functions

### PR DESCRIPTION
**Pull request by lshw54**
### Context

This pull request addresses issues related to file encoding and error handling in configuration file loading functions (`load_and_validate_config_file` and `load_and_validate_fixer_config_file`) and the HTML output generation function (`fill_html_overview_statistics`). The changes ensure that files are read and written with the correct encoding, preventing `UnicodeDecodeError`.

### Description

1. Encoding Handling:

- Updated `load_and_validate_config_file`, `load_and_validate_fixer_config_file`, and `fill_html_overview_statistics` to open files with `encoding='utf-8'` to ensure they are read and written correctly as UTF-8 encoded text.

- This change prevents `UnicodeDecodeError` caused by incorrect or unsupported file encodings.

2. Specific Exception Handling:

- Added specific exception handling for `FileNotFoundError`, `yaml.YAMLError`, and `UnicodeDecodeError` in the configuration file loading functions.
- Added specific exception handling for `FileNotFoundError` and `UnicodeDecodeError` in the HTML output generation function.
- This provides clearer error messages and makes debugging easier.

3. Default Values:

- Used default values in `stats.get()` in `fill_html_overview_statistics` to handle cases where keys might be missing.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/config/config.py [hunk 1]
```diff
@@ -99,52 +99,72 @@ def check_current_version():
...
+    except FileNotFoundError:
+        logger.critical(f"FileNotFoundError: The config file {fixer_config_file_path} was not found.")
+    except yaml.YAMLError as e:
+        logger.critical(f"YAMLError: Error parsing the YAML config file: {e}")
+    except UnicodeDecodeError as e:
+        logger.critical(f"UnicodeDecodeError: Error decoding the file {fixer_config_file_path}: {e}")
+    except Exception as e:
+        logger.critical(f"{e.__class__.__name__}[{e.__traceback__.tb_lineno}] -- {e}")
+    
+    sys.exit(1)
```
Comment (pedrooot) [inline prowler/config/config.py]: I think the `sys.exit(1)` should happen inside each exception right?

File: prowler/lib/outputs/html/html.py [hunk 2]
```diff
@@ -173,33 +173,34 @@ def fill_html(file_descriptor, finding):
...
-        )
-        sys.exit(1)
+    except FileNotFoundError:
+        logger.critical(f"FileNotFoundError: The file {filename} was not found.")
+    except UnicodeDecodeError as e:
+        logger.critical(f"UnicodeDecodeError: Error decoding the file {filename}: {e}")
+    except Exception as e:
+        logger.critical(f"{e.__class__.__name__}[{e.__traceback__.tb_lineno}] -- {e}")
+    
+    sys.exit(1)
```
Comment (pedrooot) [inline prowler/lib/outputs/html/html.py]: Same here

**Review by pedrooot — COMMENTED** [review]
Thanks for the improvements! @lshw54 Could you please review my comments? 🚀

**Comment by lshw54** [comment]
> Thanks for the improvements! @lshw54 Could you please review my comments? 🚀

Thanks!!I have made the changes to move `sys.exit(1)` inside each specific exception block in `load_and_validate_config_file`, `load_and_validate_fixer_config_file`, and `fill_html_overview_statistics`.

**Comment by pedrooot** [comment]
Could you have a look to the tests? It seems that flake8 it's failing. Using pre-commit should be enough https://docs.prowler.com/projects/prowler-open-source/en/latest/developer-guide/introduction/

**Comment by lshw54** [comment]
> Could you have a look to the tests? It seems that flake8 it's failing. Using pre-commit should be enough https://docs.prowler.com/projects/prowler-open-source/en/latest/developer-guide/introduction/

All pre-commit run result is passed

**Comment by lshw54** [comment]
Hi @pedrooot Thanks for your comment, it seems that I added some new comments and optimized a small portion of the code which resulted in a drop in coverage.

**Comment by pedrooot** [comment]
True, don't worry about that. Btw we are reviewing it now 😄

**Review by MrCloudSec — CHANGES_REQUESTED** [review]
Hi @lshw54, thanks for this fix!
Could you please change the log level to error and not exiting in those new errors? I don't think it could be a reason to exit Prowler execution since the scan can still continue.

**Comment by lshw54** [comment]
> Could you please change the log level to error and not exiting in those new errors? I don't think it could be a reason to exit Prowler execution since the scan can still continue.

I have made the requested changes. The log level has been changed to `error` for the specific exceptions, and the program will no longer exit. This allows the scan to continue even if these errors occur. Errors are now logged at the `error` level, and the function allows the scan to proceed without exiting the Prowler.

File: prowler/config/config.py [hunk 3]
```diff
@@ -99,52 +98,86 @@ def check_current_version():
...
+    except FileNotFoundError:
+        logger.error(
+            f"FileNotFoundError: The config file {fixer_config_file_path} was not found."
+        )
+    except yaml.YAMLError as error:
+        logger.error(f"YAMLError: Error parsing the YAML config file: {error}")
+    except UnicodeDecodeError as error:
+        logger.error(
+            f"UnicodeDecodeError: Error decoding the file {fixer_config_file_path}: {error}"
+        )
```
Comment (MrCloudSec) [inline prowler/config/config.py]:
```suggestion
    except FileNotFoundError:
        logger.error(
            f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}"
        )
    except yaml.YAMLError as error:
        logger.error(
            f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}"
        )
    except UnicodeDecodeError as error:
        logger.error(
            f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}"
        )
```

File: prowler/lib/outputs/html/html.py [hunk 4]
```diff
@@ -173,33 +173,38 @@ def fill_html(file_descriptor, finding):
...
+
             # Write file
-            with open(filename, "w") as file:
+            with open(filename, "w", encoding="utf-8") as file:
                 file.write(filedata)
 
+    except FileNotFoundError:
+        logger.error(f"FileNotFoundError: The file {filename} was not found.")
+    except UnicodeDecodeError as error:
+        logger.error(f"UnicodeDecodeError: Error decoding the file {filename}: {error}")
```
Comment (MrCloudSec) [inline prowler/lib/outputs/html/html.py]:
```suggestion
    except FileNotFoundError:
        logger.error(
            f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}"
        )
    except UnicodeDecodeError as error:
        logger.error(
            f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}"
        )
```

File: prowler/config/config.py [hunk 5]
```diff
@@ -99,52 +98,86 @@ def check_current_version():
...
+    except FileNotFoundError:
+        logger.error(
+            f"FileNotFoundError: The config file {config_file_path} was not found."
+        )
+    except yaml.YAMLError as error:
+        logger.error(f"YAMLError: Error parsing the YAML config file: {error}")
+    except UnicodeDecodeError as error:
+        logger.error(
+            f"UnicodeDecodeError: Error decoding the file {config_file_path}: {error}"
+        )
```
Comment (MrCloudSec) [inline prowler/config/config.py]:
```suggestion
    except FileNotFoundError:
        logger.error(
            f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}"
        )
    except yaml.YAMLError as error:
        logger.error(
            f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}"
        )
    except UnicodeDecodeError as error:
        logger.error(
            f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}"
        )
```

File: prowler/config/config.py [hunk 6]
```diff
@@ -99,52 +98,86 @@ def check_current_version():
...
+
+    Args:
+        provider (str): The provider name (e.g., 'aws', 'gcp', 'azure', 'kubernetes').
+        config_file_path (str): The path to the configuration file.
+
+    Returns:
+        dict: The configuration dictionary for the specified provider.
+
+    Raises:
+        SystemExit: If there is an error reading or parsing the configuration file.
```
Comment (MrCloudSec) [inline prowler/config/config.py]:
```suggestion
```

**Review by MrCloudSec — CHANGES_REQUESTED** [review]

**Comment by MrCloudSec** [comment]
> > Could you please change the log level to error and not exiting in those new errors? I don't think it could be a reason to exit Prowler execution since the scan can still continue.
> 
> I have made the requested changes. The log level has been changed to `error` for the specific exceptions, and the program will no longer exit. This allows the scan to continue even if these errors occur. Errors are now logged at the `error` level, and the function allows the scan to proceed without exiting the Prowler.

Thanks for the changes @lshw54 ! I also change the error messages to follow our pattern.

**Review by MrCloudSec — APPROVED** [review]

**Pull request merged**